src/Controller/Component/AuthComponent.php

```diff
@@ -745,25 +764,27 @@ protected function _getUser()
      */
     public function redirectUrl($url = null)
     {
-        if ($url !== null) {
-            $redir = $url;
-            $this->storage()->redirectUrl($redir);
-        } elseif ($redir = $this->storage()->redirectUrl()) {
-            $this->storage()->redirectUrl(false);
+        $redirectUrl = $this->request->query(static::QUERY_STRING_REDIRECT);
+        if ($redirectUrl && (substr($redirectUrl, 0, 1) !== '/')) {
+            $redirectUrl = null;
+        }
```

Comment:
We should probably fall back to reading from the session so that people can have seamless experiences when they roll out their upgraded app.